src/index.ts

```diff
@@ -234,21 +239,43 @@ QuickSQLite.transaction = (
 
   // Local transaction context object implementation
   const execute = (query: string, params?: any[]): QueryResult => {
+    if (locks[dbName].isFinalized) {
+      throw Error(
+        `Quick SQLite Error: Cannot execute query on finalized transaction: ${dbName}`
+      );
+    }
     return QuickSQLite.execute(dbName, query, params);
   };
 
+  const commit = () => {
+    const result = QuickSQLite.execute(dbName, 'COMMIT');
+    locks[dbName].isFinalized = true;
+    return result;
+  };
+
+  const rollback = () => {
+    const result = QuickSQLite.execute(dbName, 'ROLLBACK');
+    locks[dbName].isFinalized = true;
+    return result;
+  };
+
   const tx: PendingTransaction = {
     start: () => {
       try {
         QuickSQLite.execute(dbName, 'BEGIN TRANSACTION');
-        callback({ execute });
+        callback({ commit, execute, rollback });
 
-        QuickSQLite.execute(dbName, 'COMMIT');
+        if (!locks[dbName].isFinalized) {
+          commit();
+        }
       } catch (e: any) {
-        QuickSQLite.execute(dbName, 'ROLLBACK');
+        if (!locks[dbName].isFinalized) {
+          rollback();
+        }
         throw e;
       } finally {
         locks[dbName].inProgress = false;
+        locks[dbName].isFinalized = true;
```

Comment:
where is the isFinalized flag set to false again? Did you write some tests on the sample app?

Reply:
🤦‍♂️ Updated in cc2dfe5.